controllers/network/service_controller.go

```diff
@@ -0,0 +1,198 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sort"
+
+	"github.com/go-logr/logr"
+	networkv1 "github.com/openstack-k8s-operators/infra-operator/apis/network/v1beta1"
+	"golang.org/x/exp/maps"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// ServiceReconciler reconciles a Service object
+type ServiceReconciler struct {
+	client.Client
+	Kclient kubernetes.Interface
+	Log     logr.Logger
+	Scheme  *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=network.openstack.org,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=network.openstack.org,resources=services/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=network.openstack.org,resources=services/finalizers,verbs=update
+// +kubebuilder:rbac:groups=network.openstack.org,resources=dnsmasqs,verbs=get;list;watch;
+// +kubebuilder:rbac:groups=network.openstack.org,resources=dnsdatas,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Service object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *ServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	// Get a list of all DNSMasqs to trigger reconcile only for these namespaces
+	dnsmasqs := &networkv1.DNSMasqList{}
+
+	listOpts := []client.ListOption{
+		client.InNamespace(req.Namespace),
+	}
+	if err := r.Client.List(ctx, dnsmasqs, listOpts...); err != nil {
+		return ctrl.Result{}, fmt.Errorf("Unable to retrieve DNSMasqList %w", err)
+	}
+
+	for _, dnsmasq := range dnsmasqs.Items {
+		dnsHosts, err := r.getServiceDNSData(ctx, dnsmasq.Namespace)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// sort entries for DNSData spec to reduce not required updates
+		sortedDNSHosts := []networkv1.DNSHost{}
+
+		keys := maps.Keys(dnsHosts)
+		sort.Strings(keys)
+
+		for _, key := range keys {
+			sortedDNSHosts = append(sortedDNSHosts, dnsHosts[key])
+		}
+
+		err = r.createOrPatchDNSData(ctx, dnsmasq.Namespace, &dnsmasq.ObjectMeta, dnsmasq.Spec.DNSDataLabelSelectorValue, sortedDNSHosts)
```

Comment:
I think you can pass the whole dnsmasq object you got from the List query above to createOrPatchDNSData so that the createOrPatchDNSData does not need to Get it again.